test/e2e/statefulset.go

```diff
@@ -247,44 +247,417 @@ var _ = framework.KubeDescribe("StatefulSet", func() {
-
 			sst.WaitForRunningAndReady(*ss.Spec.Replicas, ss)
+			ss = sst.WaitForStatus(ss)
+			currentRevision, updateRevision := ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			Expect(currentRevision).To(Equal(updateRevision),
+				fmt.Sprintf("StatefulSet %s/%s created with update revision %s not equal to current revision %s",
+					ss.Namespace, ss.Name, updateRevision, currentRevision))
+			pods := sst.GetPodList(ss)
+			for i := range pods.Items {
+				Expect(pods.Items[i].Labels[apps.StatefulSetRevisionLabel]).To(Equal(currentRevision),
+					fmt.Sprintf("Pod %s/%s revision %s is not equal to current revision %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Labels[apps.StatefulSetRevisionLabel],
+						currentRevision))
+			}
+			sst.SortStatefulPods(pods)
+			sst.BreakPodProbe(ss, &pods.Items[1], testProbe)
+			Expect(err).NotTo(HaveOccurred())
+			ss, pods = sst.WaitForPodNotReady(ss, pods.Items[1].Name)
+			newImage := newNginxImage
+			oldImage := ss.Spec.Template.Spec.Containers[0].Image
 
+			By(fmt.Sprintf("Updating StatefulSet template: update image from %s to %s", oldImage, newImage))
+			Expect(oldImage).NotTo(Equal(newImage), "Incorrect test setup: should update to a different image")
+			ss, err = framework.UpdateStatefulSetWithRetries(c, ns, ss.Name, func(update *apps.StatefulSet) {
+				update.Spec.Template.Spec.Containers[0].Image = newImage
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Creating a new revision")
+			ss = sst.WaitForStatus(ss)
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			Expect(currentRevision).NotTo(Equal(updateRevision),
+				"Current revision should not equal update revision during rolling update")
+
+			By("Updating Pods in reverse ordinal order")
+			pods = sst.GetPodList(ss)
+			sst.SortStatefulPods(pods)
+			sst.RestorePodProbe(ss, &pods.Items[1], testProbe)
+			ss, pods = sst.WaitForPodReady(ss, pods.Items[1].Name)
+			ss, pods = sst.WaitForRollingUpdate(ss)
+			Expect(ss.Status.CurrentRevision).To(Equal(updateRevision),
+				fmt.Sprintf("StatefulSet %s/%s current revision %s does not equal updste revision %s on update completion",
+					ss.Namespace,
+					ss.Name,
+					ss.Status.CurrentRevision,
+					updateRevision))
+			for i := range pods.Items {
+				Expect(pods.Items[i].Spec.Containers[0].Image).To(Equal(newImage),
+					fmt.Sprintf(" Pod %s/%s has image %s not have new image %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Spec.Containers[0].Image,
+						newImage))
+				Expect(pods.Items[i].Labels[apps.StatefulSetRevisionLabel]).To(Equal(updateRevision),
+					fmt.Sprintf("Pod %s/%s revision %s is not equal to update revision %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Labels[apps.StatefulSetRevisionLabel],
+						updateRevision))
+			}
+
+			By("Rolling back to a previous revision")
```

Comment:
Rolling back is supported. This test ensures that the controller correctly detects a rollback and selects the previous revision instead of creating a new one.